File: hermes/watcher_core.py
"""Pure, deterministic Python port of openclaw/watcher-core.ts.

conversation-watcher (IYENTeam) is a DESIGN REFERENCE only; this is a native
reimplementation that stays in lockstep with the TypeScript core via the shared
golden fixtures (tests/fixtures/watcher-golden.json). The pinned tokenizer,
stop-word set, similarity math, and epsilon below are the parity contract:
changing them requires updating openclaw/watcher-core.ts and the fixtures
together.

No I/O, no LLM, no side effects. Targets Python 3.9+.
"""

from __future__ import annotations

import re
from typing import Dict, List, Optional, Sequence

DEFAULT_NOW = "2026-06-16T00:00:00.000Z"

PRIORITY_STACK = (
    "latest_explicit_instruction",
    "active_nudge",
    "current_topic",
    "older_summary",
)

# Parity contract: identical to STOP_WORDS in openclaw/watcher-core.ts.
STOP_WORDS = frozenset(
    {
        "the", "a", "an", "and", "or", "but", "if", "then", "so", "to", "of", "in",
        "on", "for", "is", "are", "was", "were", "be", "been", "it", "this", "that",
        "i", "you", "we", "they", "as", "at", "by", "with", "about",
        "은", "는", "이", "가", "을", "를", "에", "의", "도", "로", "으로", "고", "과", "와",
    }
)

SIMILARITY_EPSILON = 1e-9

DEFAULT_WINDOW_N = 8
DEFAULT_PERSISTENCE_K = 3
DEFAULT_SIM_THRESHOLD = 0.6
DEFAULT_PERSISTENCE_SIM_FLOOR = 0.4

_INSTRUCTION_KEYWORDS = [
    "don't", "do not", "stop", "instead", "rather",
    "아니", "하지마", "하지 마", "말고", "대신", "그만",
]

# Unicode letter/number runs, mirroring the JS /[\p{L}\p{N}]+/gu tokenizer.
# In Python 3, \w is unicode-aware for str; [^\W_] = word char excluding "_",
# i.e. letters + digits across scripts (matching \p{L}\p{N}).
_TOKEN_RE = re.compile(r"[^\W_]+", re.UNICODE)


def compact(text: str, max_len: int = 220) -> str:
    one_line = re.sub(r"\s+", " ", text).strip()
    if len(one_line) <= max_len:
        return one_line
    return one_line[: max_len - 1] + "…"


def has_any(text: str, needles: Sequence[str]) -> bool:
    lowered = text.lower()
    return any(needle.lower() in lowered for needle in needles)


def tokenize(text: str) -> List[str]:
    return [token for token in _TOKEN_RE.findall(text.lower()) if token not in STOP_WORDS]


def infer_topic(text: str) -> str:
    tokens = tokenize(text)
    if not tokens:
        return "conversation"
    return " ".join(tokens[:4])


def latest_instruction(text: str) -> Optional[str]:
    if has_any(text, _INSTRUCTION_KEYWORDS):
        return compact(text, 160)
    return None


def check(verdict: str, reason: str) -> Dict[str, str]:
    return {"verdict": verdict, "reason": reason}


def approx_gte(value: float, threshold: float, epsilon: float = SIMILARITY_EPSILON) -> bool:
    return value >= threshold - epsilon


def jaccard(a: Sequence[str], b: Sequence[str]) -> float:
    set_a = set(a)
    set_b = set(b)
    intersection = len(set_a & set_b)
    union = len(set_a) + len(set_b) - intersection
    return 0.0 if union == 0 else intersection / union


def bigrams(tokens: Sequence[str]) -> List[str]:
    return [f"{tokens[i - 1]} {tokens[i]}" for i in range(1, len(tokens))]


def similarity(a_text: str, b_text: str) -> float:
    a_tokens = tokenize(a_text)
    b_tokens = tokenize(b_text)
    token_sim = jaccard(a_tokens, b_tokens)
    gram_sim = jaccard(bigrams(a_tokens), bigrams(b_tokens))
    return max(token_sim, gram_sim)


def max_pairwise_similarity(texts: Sequence[str]) -> float:
    max_sim = 0.0
    for i in range(len(texts)):
        for j in range(i + 1, len(texts)):
            sim = similarity(texts[i], texts[j])
            if sim > max_sim:
                max_sim = sim
    return max_sim


def trailing_topic_run(topics: Sequence[str]) -> int:
    if not topics:
        return 0
    run = 1
    for i in range(len(topics) - 1, 0, -1):
        if topics[i] == topics[i - 1]:
            run += 1
        else:
            break
    return run


def _options(options: Optional[Dict[str, object]]) -> Dict[str, object]:
    return options or {}


def detect_stale_repetition(
    messages: Sequence[Dict[str, object]],
    scope_id: str,
    options: Optional[Dict[str, object]] = None,
) -> List[Dict[str, object]]:
    opts = _options(options)
    window_n = int(opts.get("windowN", DEFAULT_WINDOW_N))
    persistence_k = int(opts.get("persistenceK", DEFAULT_PERSISTENCE_K))
    sim_threshold = float(opts.get("simThreshold", DEFAULT_SIM_THRESHOLD))
    floor = float(opts.get("persistenceSimFloor", DEFAULT_PERSISTENCE_SIM_FLOOR))
    now = str(opts.get("now", DEFAULT_NOW))

    window = [m for m in messages if m.get("senderRole") == "agent"][-window_n:]
    if len(window) < 2:
        return []

    max_sim = max_pairwise_similarity([str(m.get("text", "")) for m in window])
    persistence = trailing_topic_run([infer_topic(str(m.get("text", ""))) for m in window])
    repetition = approx_gte(max_sim, sim_threshold)
    stuck = persistence >= persistence_k and approx_gte(max_sim, floor)
    if not repetition and not stuck:
        return []

    last = window[-1]
    stale_frame = infer_topic(str(last.get("text", "")))
    return [
        {
            "schema": "conversation_watcher.internal_anti_fixation_signal.v1",
            "signalId": f"sig-stale-{last.get('id')}",
            "scopeId": scope_id,
            "reason": (
                "Agent is restating near-duplicate content across recent turns."
                if repetition
                else "Agent has stayed on the same frame across consecutive turns."
            ),
            "staleFrame": stale_frame,
            "newContextEvidence": (
                "near-duplicate restatement across recent agent turns"
                if repetition
                else f"same frame across {persistence} consecutive agent turns"
            ),
            "suggestedPivot": f'Move past "{stale_frame}" and offer a genuinely fresh angle.',
            "sourceMessageIds": [m.get("id") for m in window],
            "confidence": 0.6,
            "severity": "high" if repetition else "medium",
            "fixationPattern": "stale_expression_repeated",
            "createdAt": now,
        }
    ]


def detect_correction_driven_fixation(
    messages: Sequence[Dict[str, object]],
    scope_id: str,
    options: Optional[Dict[str, object]] = None,
) -> List[Dict[str, object]]:
    opts = _options(options)
    now = str(opts.get("now", DEFAULT_NOW))
    signals: List[Dict[str, object]] = []
    for i in range(2, len(messages)):
        prev_agent = messages[i - 2]
        correction = messages[i - 1]
        current_agent = messages[i]
        if (
            prev_agent.get("senderRole") != "agent"
            or correction.get("senderRole") != "user"
            or current_agent.get("senderRole") != "agent"
        ):
            continue
        corrected = latest_instruction(str(correction.get("text", ""))) is not None
        repeated_frame = infer_topic(str(prev_agent.get("text", ""))) == infer_topic(
            str(current_agent.get("text", ""))
        )
        new_topic = infer_topic(str(correction.get("text", "")))
        if corrected and repeated_frame and infer_topic(str(current_agent.get("text", ""))) != new_topic:
            signals.append(
                {
                    "schema": "conversation_watcher.internal_anti_fixation_signal.v1",
                    "signalId": f"sig-correction-{current_agent.get('id')}",
                    "scopeId": scope_id,
                    "reason": "Agent repeated the previous frame after a newer explicit instruction.",
                    "staleFrame": infer_topic(str(prev_agent.get("text", ""))),
                    "newContextEvidence": f"{correction.get('id')}: {compact(str(correction.get('text', '')), 120)}",
                    "suggestedPivot": f"Answer from the newer frame: {new_topic}.",
                    "sourceMessageIds": [prev_agent.get("id"), correction.get("id"), current_agent.get("id")],
                    "confidence": 0.9,
                    "severity": "high",
                    "fixationPattern": "new_context_ignored_previous_frame_repeated",
                    "createdAt": now,
                }
            )
    return signals


def evaluate_fixation(
    messages: Sequence[Dict[str, object]],
    scope_id: str,
    options: Optional[Dict[str, object]] = None,
) -> List[Dict[str, object]]:
    return detect_stale_repetition(messages, scope_id, options) + detect_correction_driven_fixation(
        messages, scope_id, options
    )


def create_neutral_conversation_context(
    scope_id: str,
    messages: Sequence[Dict[str, object]],
    now: str = DEFAULT_NOW,
) -> Dict[str, object]:
    users = [m for m in messages if m.get("senderRole") == "user"]
    last_user = users[-1] if users else None
    instruction: Optional[str] = None
    for message in reversed(users):
        candidate = latest_instruction(str(message.get("text", "")))
        if candidate:
            instruction = candidate
            break
    discontinuities: List[Dict[str, str]] = []
    for i in range(1, len(users)):
        prev = users[i - 1]
        cur = users[i]
        prev_topic = infer_topic(str(prev.get("text", "")))
        cur_topic = infer_topic(str(cur.get("text", "")))
        if prev_topic != cur_topic:
            discontinuities.append(
                {
                    "fromMessageId": str(prev.get("id")),
                    "toMessageId": str(cur.get("id")),
                    "kind": "intent" if latest_instruction(str(cur.get("text", ""))) else "topic",
                    "summary": f'Context moved from "{prev_topic}" to "{cur_topic}".',
                }
            )
    return {
        "schema": "conversation_watcher.neutral_context.v1",
        "scopeId": scope_id,
        "sourceMessageIds": [m.get("id") for m in messages],
        "currentTopic": infer_topic(str(last_user.get("text", ""))) if last_user else "conversation",
        "latestExplicitInstruction": instruction,
        "recentUserIntent": compact(str(last_user.get("text", "")), 120) if last_user else None,
        "openQuestions": [
            compact(str(m.get("text", "")), 120)
            for m in users
            if re.search(r"[?？까]\s*$", str(m.get("text", "")).strip())
        ],
        "contextDiscontinuities": discontinuities,
        "summary": compact(
            " | ".join(f"{m.get('senderRole')}: {m.get('text')}" for m in messages), 280
        ),
        "confidence": 0.82 if messages else 0.1,
        "createdAt": now,
    }


def plan_external_nudge(
    runtime: str,
    signal: Dict[str, object],
    text: str,
    target: Optional[Dict[str, str]] = None,
    now: str = DEFAULT_NOW,
) -> Dict[str, object]:
    resolved_target: Dict[str, object] = {"runtime": runtime}
    if target:
        resolved_target.update(target)
    return {
        "schema": "conversation_watcher.external_nudge.v1",
        "nudgeId": f"nudge-{signal.get('signalId')}",
        "scopeId": signal.get("scopeId"),
        "target": resolved_target,
        "text": text,
        "whyNow": signal.get("reason"),
        "suggestedPivot": signal.get("suggestedPivot"),
        "sourceMessageIds": signal.get("sourceMessageIds"),
        "internalSignalId": signal.get("signalId"),
        "identityDisclosure": "agent_explicit",
        "createdAt": now,
    }


def evaluate_host_policy_gate(gate_input: Dict[str, object]) -> Dict[str, object]:
    signal = gate_input["signal"]
    source_thread = gate_input.get("sourceThreadId")
    target_thread = gate_input.get("targetThreadId")
    cross_thread = bool(gate_input.get("crossThreadRisk"))
    thread_matches = not source_thread or not target_thread or source_thread == target_thread
    thread_ok = thread_matches and not cross_thread
    if cross_thread:
        thread_reason = "signal sources span multiple threads"
    elif thread_matches:
        thread_reason = "thread matches or is unspecified"
    else:
        thread_reason = "source and target thread mismatch"

    if gate_input.get("shadowMode"):
        suppressed_reason: Optional[str] = "shadow_mode"
    elif gate_input.get("cooldownHit"):
        suppressed_reason = "cooldown"
    elif gate_input.get("duplicateHit"):
        suppressed_reason = "duplicate"
    elif gate_input.get("privacyRisk"):
        suppressed_reason = "privacy"
    elif not thread_ok:
        suppressed_reason = "thread_mismatch"
    else:
        suppressed_reason = None
    allowed = suppressed_reason is None

    return {
        "schema": "conversation_watcher.host_policy_gate_audit.v1",
        "runtime": gate_input["runtime"],
        "allowed": allowed,
        "reason": "host policy gate allowed native delivery"
        if allowed
        else f"suppressed: {suppressed_reason}",
        "threadId": target_thread,
        "sessionId": gate_input.get("sessionId"),
        "sourceMessageIds": signal.get("sourceMessageIds"),
        "cooldownKey": f"{signal.get('scopeId')}:{signal.get('fixationPattern')}",
        "duplicateCheck": check(
            "blocked" if gate_input.get("duplicateHit") else "passed",
            "duplicate nudge" if gate_input.get("duplicateHit") else "no duplicate",
        ),
        "privacyCheck": check(
            "blocked" if gate_input.get("privacyRisk") else "passed",
            "privacy risk" if gate_input.get("privacyRisk") else "no privacy risk",
        ),
        "threadCheck": check("passed" if thread_ok else "blocked", thread_reason),
        "criticConfidence": gate_input.get("criticConfidence"),
        "fixationPattern": signal.get("fixationPattern"),
        "internalSignalId": signal.get("signalId"),
        "deliveryMessageId": gate_input.get("deliveryMessageId") if allowed else None,
        "suppressedReason": suppressed_reason,
        "createdAt": gate_input.get("now", DEFAULT_NOW),
    }


def evaluate_fixture_result(
    fixture: Dict[str, object],
    options: Optional[Dict[str, object]] = None,
) -> Dict[str, object]:
    signals = evaluate_fixation(
        fixture.get("rawMessages", []), str(fixture.get("scopeId", "")), options
    )
    return {
        "fixated": len(signals) > 0,
        "patterns": [s.get("fixationPattern") for s in signals],
    }

File: openclaw/watcher-core.ts
// Pure, side-effect-free reimplementation of the conversation-watcher
// anti-fixation core for hent-ai. conversation-watcher (IYENTeam) is a DESIGN
// REFERENCE only — no external dependency is consumed. See the approved plan
// .gjc/plans/ralplan/2026-06-18-0155-3fca/stage-03-final.md.
//
// This module is deterministic and has no I/O and no LLM calls so it can be
// covered to 100% and kept in lockstep with the Python port (hermes/watcher_core.py)
// via the shared golden fixtures. The pinned tokenizer / stop-word list /
// epsilon comparison below are the parity contract: changing them requires
// updating both surfaces and the golden fixtures together.

export const DEFAULT_NOW = "2026-06-16T00:00:00.000Z";

export const PRIORITY_STACK = [
  "latest_explicit_instruction",
  "active_nudge",
  "current_topic",
  "older_summary",
] as const;
export type PriorityStackEntry = (typeof PRIORITY_STACK)[number];

export type RuntimeName = "openclaw" | "hermes";
export type SenderRole = "user" | "agent" | "system";
export type GateCheckVerdict = "passed" | "blocked" | "not_applicable";
export type FixationPattern =
  | "stale_expression_repeated"
  | "new_context_ignored_previous_frame_repeated";

export interface RawConversationMessage {
  id: string;
  senderRole: SenderRole;
  ts: string;
  text: string;
  threadId?: string;
  sessionId?: string;
  metadata?: Record<string, unknown>;
}

export type NudgeKind =
  | "tone"
  | "angle"
  | "hypothesis"
  | "priority"
  | "constraint"
  | "exploration"
  | "format";
export type NudgeStrength = "weak" | "normal" | "strong";

/**
 * A soft, in-conversation steering hint derived from the neutral context. Not
 * delivered directly; it informs the prompt-context priority stack. Retained in
 * the core for parity with the reference schema and for prompt-context work that
 * lands with the adapters (see PRIORITY_STACK).
 */
export interface ContextNudge {
  schema: "conversation_watcher.context_nudge.v1";
  scopeId: string;
  sourceMessageId: string;
  authorId?: string;
  kind: NudgeKind;
  instruction: string;
  strength: NudgeStrength;
  lifetime: { mode: "turns" | "messages" | "minutes" | "until_replaced"; value: number | null };
  priority: number;
  safetyBoundary: "soft_steering_not_policy_override";
  createdAt: string;
  expiresAt: string | null;
}

export interface ConversationDiscontinuity {
  fromMessageId: string;
  toMessageId: string;
  kind: "topic" | "intent";
  summary: string;
}

export interface NeutralConversationContext {
  schema: "conversation_watcher.neutral_context.v1";
  scopeId: string;
  sourceMessageIds: string[];
  currentTopic: string;
  latestExplicitInstruction: string | null;
  recentUserIntent: string | null;
  openQuestions: string[];
  contextDiscontinuities: ConversationDiscontinuity[];
  summary: string;
  confidence: number;
  createdAt: string;
}

export interface InternalAntiFixationSignal {
  schema: "conversation_watcher.internal_anti_fixation_signal.v1";
  signalId: string;
  scopeId: string;
  reason: string;
  staleFrame: string;
  newContextEvidence: string;
  suggestedPivot: string;
  sourceMessageIds: string[];
  confidence: number;
  severity: "low" | "medium" | "high";
  fixationPattern: FixationPattern;
  createdAt: string;
}

export interface ExternalNudge {
  schema: "conversation_watcher.external_nudge.v1";
  nudgeId: string;
  scopeId: string;
  target: { runtime: RuntimeName; channel?: string; threadId?: string; sessionId?: string };
  text: string;
  whyNow: string;
  suggestedPivot: string;
  sourceMessageIds: string[];
  internalSignalId: string;
  identityDisclosure: "agent_explicit";
  createdAt: string;
}

export interface GateCheckResult {
  verdict: GateCheckVerdict;
  reason: string;
}

export interface HostPolicyGateAudit {
  schema: "conversation_watcher.host_policy_gate_audit.v1";
  runtime: RuntimeName;
  allowed: boolean;
  reason: string;
  threadId?: string;
  sessionId?: string;
  sourceMessageIds: string[];
  cooldownKey: string;
  duplicateCheck: GateCheckResult;
  privacyCheck: GateCheckResult;
  threadCheck: GateCheckResult;
  criticConfidence: number;
  fixationPattern: FixationPattern;
  internalSignalId: string;
  deliveryMessageId?: string;
  suppressedReason?: string;
  createdAt: string;
}

export interface GoldenReplayFixture {
  name: string;
  description: string;
  scopeId: string;
  rawMessages: RawConversationMessage[];
  expectedFixated: boolean;
  expectedPatterns: FixationPattern[];
}

// --------------------------------------------------------------------------
// Pinned deterministic helpers (parity contract with hermes/watcher_core.py)
// --------------------------------------------------------------------------

export const STOP_WORDS: ReadonlySet<string> = new Set<string>([
  "the", "a", "an", "and", "or", "but", "if", "then", "so", "to", "of", "in",
  "on", "for", "is", "are", "was", "were", "be", "been", "it", "this", "that",
  "i", "you", "we", "they", "as", "at", "by", "with", "about",
  "은", "는", "이", "가", "을", "를", "에", "의", "도", "로", "으로", "고", "과", "와",
]);

export const SIMILARITY_EPSILON = 1e-9;

export function compact(text: string, max = 220): string {
  const oneLine = text.replace(/\s+/g, " ").trim();
  return oneLine.length <= max ? oneLine : `${oneLine.slice(0, max - 1)}…`;
}

export function hasAny(text: string, needles: string[]): boolean {
  const lowered = text.toLowerCase();
  return needles.some((needle) => lowered.includes(needle.toLowerCase()));
}

/** Pinned tokenizer: lowercase, keep unicode letter/number runs, drop stop words. */
export function tokenize(text: string): string[] {
  const matched = text.toLowerCase().match(/[\p{L}\p{N}]+/gu) ?? [];
  return matched.filter((token) => !STOP_WORDS.has(token));
}

export function inferTopic(text: string): string {
  const tokens = tokenize(text);
  if (tokens.length === 0) return "conversation";
  return tokens.slice(0, 4).join(" ");
}

export function latestInstruction(text: string): string | null {
  if (
    hasAny(text, [
      "don't", "do not", "stop", "instead", "rather",
      "아니", "하지마", "하지 마", "말고", "대신", "그만",
    ])
  ) {
    return compact(text, 160);
  }
  return null;
}

export function check(verdict: GateCheckVerdict, reason: string): GateCheckResult {
  return { verdict, reason };
}

export function approxGte(value: number, threshold: number, epsilon = SIMILARITY_EPSILON): boolean {
  return value >= threshold - epsilon;
}

export function jaccard(a: string[], b: string[]): number {
  const setA = new Set(a);
  const setB = new Set(b);
  let intersection = 0;
  for (const token of setA) {
    if (setB.has(token)) intersection += 1;
  }
  const union = setA.size + setB.size - intersection;
  return union === 0 ? 0 : intersection / union;
}

export function bigrams(tokens: string[]): string[] {
  const grams: string[] = [];
  for (let i = 1; i < tokens.length; i += 1) {
    grams.push(`${tokens[i - 1]} ${tokens[i]}`);
  }
  return grams;
}

/** Max of token-set Jaccard and bigram-set Jaccard between two texts. */
export function similarity(aText: string, bText: string): number {
  const aTokens = tokenize(aText);
  const bTokens = tokenize(bText);
  const tokenSim = jaccard(aTokens, bTokens);
  const gramSim = jaccard(bigrams(aTokens), bigrams(bTokens));
  return Math.max(tokenSim, gramSim);
}

export function maxPairwiseSimilarity(texts: string[]): number {
  let max = 0;
  for (let i = 0; i < texts.length; i += 1) {
    for (let j = i + 1; j < texts.length; j += 1) {
      const sim = similarity(texts[i]!, texts[j]!);
      if (sim > max) max = sim;
    }
  }
  return max;
}

export function trailingTopicRun(topics: string[]): number {
  if (topics.length === 0) return 0;
  let run = 1;
  for (let i = topics.length - 1; i > 0; i -= 1) {
    if (topics[i] === topics[i - 1]) run += 1;
    else break;
  }
  return run;
}

// --------------------------------------------------------------------------
// Detection
// --------------------------------------------------------------------------

export interface FixationOptions {
  windowN?: number;
  persistenceK?: number;
  simThreshold?: number;
  persistenceSimFloor?: number;
  now?: string;
}

export const DEFAULT_WINDOW_N = 8;
export const DEFAULT_PERSISTENCE_K = 3;
export const DEFAULT_SIM_THRESHOLD = 0.6;
export const DEFAULT_PERSISTENCE_SIM_FLOOR = 0.4;

/**
 * High-recall self-repetition prefilter. Fires when the agent restates
 * near-duplicate content across recent turns (no user correction required),
 * OR stays stuck on the same frame for several moderately-similar turns. It is
 * intentionally permissive; the downstream LLM critic is the precision gate.
 */
export function detectStaleRepetition(
  messages: RawConversationMessage[],
  scopeId: string,
  options: FixationOptions = {},
): InternalAntiFixationSignal[] {
  const windowN = options.windowN ?? DEFAULT_WINDOW_N;
  const persistenceK = options.persistenceK ?? DEFAULT_PERSISTENCE_K;
  const simThreshold = options.simThreshold ?? DEFAULT_SIM_THRESHOLD;
  const floor = options.persistenceSimFloor ?? DEFAULT_PERSISTENCE_SIM_FLOOR;
  const now = options.now ?? DEFAULT_NOW;

  const window = messages.filter((m) => m.senderRole === "agent").slice(-windowN);
  if (window.length < 2) return [];

  const maxSim = maxPairwiseSimilarity(window.map((m) => m.text));
  const persistence = trailingTopicRun(window.map((m) => inferTopic(m.text)));
  const repetition = approxGte(maxSim, simThreshold);
  const stuck = persistence >= persistenceK && approxGte(maxSim, floor);
  if (!repetition && !stuck) return [];

  const last = window[window.length - 1]!;
  const staleFrame = inferTopic(last.text);
  return [
    {
      schema: "conversation_watcher.internal_anti_fixation_signal.v1",
      signalId: `sig-stale-${last.id}`,
      scopeId,
      reason: repetition
        ? "Agent is restating near-duplicate content across recent turns."
        : "Agent has stayed on the same frame across consecutive turns.",
      staleFrame,
      newContextEvidence: repetition
        ? "near-duplicate restatement across recent agent turns"
        : `same frame across ${persistence} consecutive agent turns`,
      suggestedPivot: `Move past "${staleFrame}" and offer a genuinely fresh angle.`,
      sourceMessageIds: window.map((m) => m.id),
      confidence: 0.6,
      severity: repetition ? "high" : "medium",
      fixationPattern: "stale_expression_repeated",
      createdAt: now,
    },
  ];
}

/**
 * Additional reference-style signal: the agent repeats the previous frame even
 * after the user issued a newer explicit instruction. Retained as a secondary
 * signal only; it is NOT the primary trigger.
 */
export function detectCorrectionDrivenFixation(
  messages: RawConversationMessage[],
  scopeId: string,
  options: FixationOptions = {},
): InternalAntiFixationSignal[] {
  const now = options.now ?? DEFAULT_NOW;
  const signals: InternalAntiFixationSignal[] = [];
  for (let i = 2; i < messages.length; i += 1) {
    const prevAgent = messages[i - 2]!;
    const correction = messages[i - 1]!;
    const currentAgent = messages[i]!;
    if (
      prevAgent.senderRole !== "agent" ||
      correction.senderRole !== "user" ||
      currentAgent.senderRole !== "agent"
    ) {
      continue;
    }
    const corrected = latestInstruction(correction.text) !== null;
    const repeatedFrame = inferTopic(prevAgent.text) === inferTopic(currentAgent.text);
    const newTopic = inferTopic(correction.text);
    if (corrected && repeatedFrame && inferTopic(currentAgent.text) !== newTopic) {
      signals.push({
        schema: "conversation_watcher.internal_anti_fixation_signal.v1",
        signalId: `sig-correction-${currentAgent.id}`,
        scopeId,
        reason: "Agent repeated the previous frame after a newer explicit instruction.",
        staleFrame: inferTopic(prevAgent.text),
        newContextEvidence: `${correction.id}: ${compact(correction.text, 120)}`,
        suggestedPivot: `Answer from the newer frame: ${newTopic}.`,
        sourceMessageIds: [prevAgent.id, correction.id, currentAgent.id],
        confidence: 0.9,
        severity: "high",
        fixationPattern: "new_context_ignored_previous_frame_repeated",
        createdAt: now,
      });
    }
  }
  return signals;
}

export function evaluateFixation(
  messages: RawConversationMessage[],
  scopeId: string,
  options: FixationOptions = {},
): InternalAntiFixationSignal[] {
  return [
    ...detectStaleRepetition(messages, scopeId, options),
    ...detectCorrectionDrivenFixation(messages, scopeId, options),
  ];
}

export function createNeutralConversationContext(
  scopeId: string,
  messages: RawConversationMessage[],
  now = DEFAULT_NOW,
): NeutralConversationContext {
  const users = messages.filter((m) => m.senderRole === "user");
  const lastUser = users[users.length - 1];
  const instruction =
    [...users].reverse().map((m) => latestInstruction(m.text)).find(Boolean) ?? null;
  const discontinuities: ConversationDiscontinuity[] = [];
  for (let i = 1; i < users.length; i += 1) {
    const prev = users[i - 1]!;
    const cur = users[i]!;
    const prevTopic = inferTopic(prev.text);
    const curTopic = inferTopic(cur.text);
    if (prevTopic !== curTopic) {
      discontinuities.push({
        fromMessageId: prev.id,
        toMessageId: cur.id,
        kind: latestInstruction(cur.text) ? "intent" : "topic",
        summary: `Context moved from "${prevTopic}" to "${curTopic}".`,
      });
    }
  }
  return {
    schema: "conversation_watcher.neutral_context.v1",
    scopeId,
    sourceMessageIds: messages.map((m) => m.id),
    currentTopic: lastUser ? inferTopic(lastUser.text) : "conversation",
    latestExplicitInstruction: instruction,
    recentUserIntent: lastUser ? compact(lastUser.text, 120) : null,
    openQuestions: users
      .filter((m) => /[?？까]\s*$/.test(m.text.trim()))
      .map((m) => compact(m.text, 120)),
    contextDiscontinuities: discontinuities,
    summary: compact(messages.map((m) => `${m.senderRole}: ${m.text}`).join(" | "), 280),
    confidence: messages.length ? 0.82 : 0.1,
    createdAt: now,
  };
}

export function planExternalNudge(
  runtime: RuntimeName,
  signal: InternalAntiFixationSignal,
  text: string,
  target: { channel?: string; threadId?: string; sessionId?: string } = {},
  now = DEFAULT_NOW,
): ExternalNudge {
  return {
    schema: "conversation_watcher.external_nudge.v1",
    nudgeId: `nudge-${signal.signalId}`,
    scopeId: signal.scopeId,
    target: { runtime, ...target },
    text,
    whyNow: signal.reason,
    suggestedPivot: signal.suggestedPivot,
    sourceMessageIds: signal.sourceMessageIds,
    internalSignalId: signal.signalId,
    identityDisclosure: "agent_explicit",
    createdAt: now,
  };
}

export interface HostPolicyGateInput {
  runtime: RuntimeName;
  signal: InternalAntiFixationSignal;
  criticConfidence: number;
  sourceThreadId?: string;
  targetThreadId?: string;
  sessionId?: string;
  shadowMode?: boolean;
  cooldownHit?: boolean;
  duplicateHit?: boolean;
  privacyRisk?: boolean;
  crossThreadRisk?: boolean;
  deliveryMessageId?: string;
  now?: string;
}

export function evaluateHostPolicyGate(input: HostPolicyGateInput): HostPolicyGateAudit {
  const threadMatches =
    !input.sourceThreadId || !input.targetThreadId || input.sourceThreadId === input.targetThreadId;
  const threadOk = threadMatches && !input.crossThreadRisk;
  const threadReason = input.crossThreadRisk
    ? "signal sources span multiple threads"
    : threadMatches
      ? "thread matches or is unspecified"
      : "source and target thread mismatch";
  const suppressedReason = input.shadowMode
    ? "shadow_mode"
    : input.cooldownHit
      ? "cooldown"
      : input.duplicateHit
        ? "duplicate"
        : input.privacyRisk
          ? "privacy"
          : !threadOk
            ? "thread_mismatch"
            : undefined;
  const allowed = suppressedReason === undefined;
  return {
    schema: "conversation_watcher.host_policy_gate_audit.v1",
    runtime: input.runtime,
    allowed,
    reason: allowed ? "host policy gate allowed native delivery" : `suppressed: ${suppressedReason}`,
    threadId: input.targetThreadId,
    sessionId: input.sessionId,
    sourceMessageIds: input.signal.sourceMessageIds,
    cooldownKey: `${input.signal.scopeId}:${input.signal.fixationPattern}`,
    duplicateCheck: check(
      input.duplicateHit ? "blocked" : "passed",
      input.duplicateHit ? "duplicate nudge" : "no duplicate",
    ),
    privacyCheck: check(
      input.privacyRisk ? "blocked" : "passed",
      input.privacyRisk ? "privacy risk" : "no privacy risk",
    ),
    threadCheck: check(threadOk ? "passed" : "blocked", threadReason),
    criticConfidence: input.criticConfidence,
    fixationPattern: input.signal.fixationPattern,
    internalSignalId: input.signal.signalId,
    deliveryMessageId: allowed ? input.deliveryMessageId : undefined,
    suppressedReason,
    createdAt: input.now ?? DEFAULT_NOW,
  };
}

/** Run the pure pipeline over a fixture and return the structured result used by parity tests. */
export function evaluateFixtureResult(
  fixture: GoldenReplayFixture,
  options: FixationOptions = {},
): { fixated: boolean; patterns: FixationPattern[] } {
  const signals = evaluateFixation(fixture.rawMessages, fixture.scopeId, options);
  return {
    fixated: signals.length > 0,
    patterns: signals.map((s) => s.fixationPattern),
  };
}
